test/first_party/CMakeLists.txt

```
@@ -0,0 +1,141 @@
+
+# First party tests
+
+# There are three options for cross-compilation of an RV32 program.
+#
+# 1. Clang, which is sensibly designed to be able to cross-compile to any architecture it supports.
+#    Unfortunately it is possible to build it without RISC-V support, e.g. RHEL 8 does this, so we
+#    need to check that. You can download a binary from https://github.com/llvm/llvm-project/releases
+# 2. GCC via riscv32-unknown-elf-gcc. Unlike Clang you have to specifically build GCC as a cross
+#    compiler. You can download a binary from https://github.com/riscv-collab/riscv-gnu-toolchain/releases
+# 3. GCC via riscv64-unknown-elf-gcc, but only if it has "multilib" support, so we'll try riscv32-... first.
+
+find_program(CLANG_BIN "clang")
+find_program(GCC_BIN_RV32 "riscv32-unknown-elf-gcc")
+find_program(GCC_BIN_RV64 "riscv64-unknown-elf-gcc")
+
+if (CLANG_BIN)
+    message(STATUS "Found clang: ${CLANG_BIN}")
+endif()
+if (GCC_BIN_RV32)
+    message(STATUS "Found riscv32-unknown-elf-gcc: ${GCC_BIN_RV32}")
+endif()
+if (GCC_BIN_RV64)
+    message(STATUS "Found riscv64-unknown-elf-gcc: ${GCC_BIN_RV64}")
+endif()
+
+set(CROSS_COMPILER_COMMAND_RV32)
+set(CROSS_COMPILER_COMMAND_RV64)
+
+# Prefer Clang.
+if (CLANG_BIN)
+    # Check it supports RISC-V.
+    execute_process(
+        COMMAND clang -print-targets
+        OUTPUT_VARIABLE clang_targets
+        COMMAND_ERROR_IS_FATAL ANY
+    )
+
+    # Check if `riscv32` is present in the output
+    if (clang_targets MATCHES "riscv32")
+        set(CROSS_COMPILER_COMMAND_RV32 ${CLANG_BIN} --target=riscv32)
+        set(CROSS_COMPILER_COMMAND_RV64 ${CLANG_BIN} --target=riscv64)
+    else()
+        message(WARNING "Your Clang compiler does not support RISC-V (see '${CLANG_BIN} -print-targets'). Please download one from https://github.com/llvm/llvm-project/releases")
+    endif()
+endif()
+
+# Prefer riscv32-unknown-elf-gcc to riscv64-unknown-elf-gcc for RV32.
+if (GCC_BIN_RV32)
+    if (NOT CROSS_COMPILER_COMMAND_RV32)
+        set(CROSS_COMPILER_COMMAND_RV32 ${GCC_BIN_RV32})
+    endif()
+endif()
+
+if (GCC_BIN_RV64)
+    if (NOT CROSS_COMPILER_COMMAND_RV32)
+        # It might support multilib.
+        set(CROSS_COMPILER_COMMAND_RV32 ${GCC_BIN_RV64})
```

```suggestion
        set(CROSS_COMPILER_COMMAND_RV32 ${GCC_BIN_RV64} -m32)
```
Isn't this needed to target RV32?

---

I believe `-m32` is specific to x86. For riscv, this is all handled by the march string.

---

-m32 works for many architectures, but you are right using -march=rv32... will have the same effect. Please ignore.